tests/unit/package_managers/bundler/test_parser.py

```diff
@@ -15,13 +16,26 @@
     GEMFILE_LOCK,
     BundlerDependency,
     GemDependency,
+    GemPlatformSpecificDependency,
     GitDependency,
     PathDependency,
     parse_lockfile,
 )
 from cachi2.core.rooted_path import RootedPath
 from tests.common_utils import GIT_REF
 
+RegexpStr = str  # a string representing a regular expression.
+
+
+def some_message_contains_substring(substring: RegexpStr, messages: Iterable[str]) -> bool:
+    """Check if substring-matching regexp could be found in any message.
+
+    This produces a bit less coupling between tests and code than
+    checking for a full message.
+    """
+    r = re.compile(substring)
+    return any(r.match(m) is not None for m in messages)
+
```

Comment:
If I replace the function with `assert "..." in caplog.text` everything passes. But it is not ideal either.

> Now imagine an error message which has been constructed at multiple places (unwinding of the exception stack), each adding some data specific bits.

This feels like a message that could be constructed at multiple places:
```
assert some_message_contains_substring("Found a binary dependency", caplog.messages)
assert some_message_contains_substring("Downloading binary dependency", caplog.messages)
```
So I don't think we have to do anything here (at least for now). 
Just use `caplog.messages` and check full log messages as we do elsewhere in the tests.

Reply:
> This feels like a message that could be constructed at multiple places:
> ```
> assert some_message_contains_substring("Found a binary dependency", caplog.messages)
> assert some_message_contains_substring("Downloading binary dependency", caplog.messages)
>```

What I meant was a message like this:
`FetchError: Failed to fetch dependencies: Failed to process file 'XYZ': 'No such file or directory'`

where the message itself consists of 3 parts consisting of 3 different streams of information or just context wrapping whatever you want to call it where the 2nd part of the msg is tied to file some operations and the 3rd part gives you the exact system error (based on errno) why the file operation failed. The messages can be arbitrarily complex and regex handling is useful in such cases.
That said, it's not a hard blocker for me, the function name isn't ideal though and I'd surely use the word `assert` in its name to signal it's a wrapper over `assert`.

Reply:
Note that our logging is far from ideal, so we may want to improve on that in the future. That is irrelevant to this thread and PR and so I won't digress.

Reply:
Oh I found this: https://github.com/containerbuildsystem/cachi2/pull/128

Reply:
> Checking messages is a questionable idea because it tightly couples code and tests. A typo or a new, clearer message would mean that tests must change too and this is bad because no change is logic resulted in a change to tests. Tests must check logic, exact shape of side effects is unimportant. Our logic demands that we emit some messages...

100% agreed! That's why I suggested regex pattern searching in the first place to avoid the need to change tests when we want to add a small additional check that we're emitting an error/exception of the right type and that the contents more or less makes sense instead of matching exact strings which I'm also not a fan of.
We're doing that already, we can improve and it's not a blocker for me if we don't do it here ultimately, I just had a suggestion on regexes and you did that. Generalizing the function can come later (if we don't forget).